### PR TITLE
[#140077551] Add "Not Found" responses when identifiers don't match

### PR DESCRIPTION
**Pull request by waggles**
* Adds reader and controller as `CardReader` and `ControlDevice` (to avoid naming confusion with MVC controllers, hopefully)
* builds on #897 , when that's in I can clean up the commits a bit

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -88,6 +88,19 @@
   add_index "bundle_products", ["bundle_product_id"], name: "fk_bundle_prod_prod", using: :btree
   add_index "bundle_products", ["product_id"], name: "fk_bundle_prod_bundle", using: :btree
 
+  create_table "card_readers", force: :cascade do |t|
```
Comment (jhanggi) [inline db/schema.rb]: I'm wondering if we should prefix these with `secure_room_`. Part of me feels we should do the model classes could be namespaced (`SecureRooms::CardReader`). We do the namespace part for Projects (the model is `Projects::Project`), which I confess is a little awkward, much like `SecureRooms::Room` would have been. What do you think?

File: db/schema.rb [hunk 2]
```diff
@@ -88,6 +88,19 @@
...
+    t.integer  "control_device_id", limit: 4, null: false
+    t.boolean  "entrance",                    null: false
+    t.datetime "created_at",                  null: false
+    t.datetime "updated_at",                  null: false
+  end
+
+  create_table "control_devices", force: :cascade do |t|
+    t.integer  "secure_room_id", limit: 4, null: false
```
Comment (jhanggi) [inline db/schema.rb]: I'd maybe like this to be genericized to `product_id`. I could go either way, though.

But we should add an index and a foreign key on it (fyi, mysql automatically creates an index when you create a foreign key, but oracle doesn't, so you'll need to explicitly `add_index` as well as adding the foreign key.

File: db/schema.rb [hunk 3]
```diff
@@ -88,6 +88,19 @@
   add_index "bundle_products", ["bundle_product_id"], name: "fk_bundle_prod_prod", using: :btree
   add_index "bundle_products", ["product_id"], name: "fk_bundle_prod_bundle", using: :btree
 
+  create_table "card_readers", force: :cascade do |t|
+    t.integer  "control_device_id", limit: 4, null: false
+    t.boolean  "entrance",                    null: false
```
Comment (jhanggi) [inline db/schema.rb]: Hmm, I'm not crazy about this name, but I can't come up with anything better. I originally thought `direction`, but then true/false doesn't really make sense with it. It's probably fine for now and we can change it if we come up with something better.

File: vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb [hunk 4]
```diff
@@ -14,4 +16,10 @@ def scan
     render json: response_json, status: :forbidden
   end
 
+  def load_models
+    @user = User.find_by!(card_number: params[:card_id])
+    @control_device = ControlDevice.find(params[:controller_id])
+    @card_reader = CardReader.find(params[:reader_id])
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb]: Since we have the foreign key, we should probably `@card_reader = @control_devise.card_readers.find` to scope it down a little better.

File: vendor/engines/secure_rooms/app/models/card_reader.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class CardReader < ActiveRecord::Base
+
+   belongs_to :control_device
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/app/models/card_reader.rb]: Rubocop is flagging a spacing issue ("Use 2 (not 3) spaces for indentation.")

File: vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb [hunk 6]
```diff
@@ -7,11 +7,44 @@
...
+           controller_id: control_device_id,
+           reader_id: card_reader_id
+    end
+
+    subject { response }
+
+    let(:card_reader) { create :card_reader }
+    let(:card_user) { create :user, card_number: '123456' }
+
+    let(:card_reader_id) { card_reader.id }
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb]: I'm not really sure these `let`s are carrying their weight. I think you'd be fine if you used the dot version in the `post`.

File: vendor/engines/secure_rooms/spec/factories/secure_room.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :secure_room, class: SecureRoom do
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/spec/factories/secure_room.rb]: I'm not sure the `class` option is necessary. Is there a way we could do `parent: :product` to get these attributes from what already exist?

File: db/schema.rb [hunk 8]
```diff
@@ -688,6 +701,7 @@
     t.datetime "reset_password_sent_at"
     t.integer  "uid",                    limit: 4
     t.datetime "deactivated_at"
+    t.string   "card_number",            limit: 255
```
Comment (jhanggi) [inline db/schema.rb]: We should probably add an index on `card_number`

File: vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb (hunk 4 above)
Comment (jhanggi) [inline vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb]: We might also need to add a custom `rescue_from` to these controllers. We don't want to send back HTML if we can avoid it. This might be a question for Dan, but should we return in the body why it's 404ing?

File: db/schema.rb (hunk 1 above)
Comment (waggles) [inline db/schema.rb]: actually yeah I prefer that

File: db/schema.rb (hunk 3 above)
Comment (waggles) [inline db/schema.rb]: haha yeah I actually could have just left it out for now, after like 30s of trying to think about what a good name was I just moved on

File: db/schema.rb (hunk 2 above)
Comment (waggles) [inline db/schema.rb]: I'm cool with `product_id` I'll update

File: db/schema.rb (hunk 1 above)
Comment (jhanggi) [inline db/schema.rb]: On second thought (aka shower thought): do we need two models? We can probably get away with a single model with a controller id and a reader id (and those might have been what you were thinking as opposed to database ids, which is how I read them). I suspect that would make it much easier to build the crud interface and then even add to a product. Having to enter the controller id twice in the interface isn't a big deal in my mind.

File: vendor/engines/secure_rooms/app/models/card_reader.rb (hunk 5 above)
Comment (waggles) [inline vendor/engines/secure_rooms/app/models/card_reader.rb]: I just realized I could probably just run rubocop on `engines/secure_rooms` and spot these much more easily 😝

File: vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb (hunk 6 above)
Comment (waggles) [inline vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb]: it just made the `context`s easier since there's no calling accidentally id on `nil`

File: db/schema.rb (hunk 1 above)
Comment (waggles) [inline db/schema.rb]: oh wow yeah I hadn't fully thought about what the id's actually are--do we have examples?

File: db/schema.rb (hunk 8 above)
Comment (waggles) [inline db/schema.rb]: Updated this over in #918 since it seemed to fit

File: db/schema.rb (hunk 1 above)
Comment (waggles) [inline db/schema.rb]: since I'm not sure yet what data is going in these, I'd like to just leave it as is until I'm sure--it should be easy enough to migrate into whatever we need at that point

File: vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb (hunk 4 above)
Comment (waggles) [inline vendor/engines/secure_rooms/app/controllers/secure_rooms_api/scans_controller.rb]: I added the error message for now and updated the README

File: vendor/engines/secure_rooms/spec/factories/secure_room.rb (hunk 7 above)
Comment (waggles) [inline vendor/engines/secure_rooms/spec/factories/secure_room.rb]: ah okay I added in the `parent:` but I think it also needs the `class:` to generate the right model

File: db/schema.rb (hunk 1 above)
Comment (jhanggi) [inline db/schema.rb]: What do you think of collapsing the two tables into one? If you really want to do that as a separate PR, that's fine. It just seems like that would be extra work as you would then need to write data migrations as opposed to just schema changes.

File: vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb (hunk 6 above)
Comment (jhanggi) [inline vendor/engines/secure_rooms/spec/controllers/secure_rooms_api/scans_controller_spec.rb]: Ok, gotcha.

**Comment by waggles** [comment]
@jhanggi oops I was pushing some updates there might be some oddities

File: db/schema.rb (hunk 1 above)
Comment (waggles) [inline db/schema.rb]: ok yeah based on chat, I'll go ahead and reduce to one table

File: vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_reader_and_control_device.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+class CreateCardReaderAndControlDevice < ActiveRecord::Migration
+
+  def change
+    create_table :secure_rooms_control_devices do |t|
+      t.references :product, null: false
+      t.foreign_key :products
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_reader_and_control_device.rb]: So funny story: Oracle doesn't automatically create an index on this (or at least it didn't the last time I tried). We can leave it out for now to see what happens when we pull it down into NU, but I wanted to call that out.

Comment (waggles) [inline vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_reader_and_control_device.rb]: oy, I continue to underestimate oracle's situation 😵 might as well add it if I'm updating here

Comment (jhanggi) [inline vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_reader_and_control_device.rb]: We can leave it out for now; I'm actually curious if it's gotten fixed recently.

**Comment by waggles** [comment]
@jhanggi ready for 👀

File: vendor/engines/secure_rooms/app/models/secure_rooms/card_reader.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+module SecureRooms
+
+  class CardReader < ActiveRecord::Base
+
+    belongs_to :secure_room, foreign_key: :product_id
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/app/models/secure_rooms/card_reader.rb]: Will validation come in the next PR (which I'm assuming will be the crud for these)? I'm cool if so, just want to check.

File: vendor/engines/secure_rooms/app/models/secure_room.rb [hunk 11]
```diff
@@ -1,3 +1,6 @@
 class SecureRoom < Product
 
+  has_many :secure_rooms_control_devices
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/app/models/secure_room.rb]: Looks like these `has_many` haven't been updated for the new model yet.

File: vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_readers.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+class CreateCardReaders < ActiveRecord::Migration
+
+  def change
+    create_table :secure_rooms_card_readers do |t|
+      t.references :product, null: false
```
Comment (jhanggi) [inline vendor/engines/secure_rooms/db/migrate/20170301120659_create_card_readers.rb]: Still need an explicit index set because Oracle.

File: vendor/engines/secure_rooms/app/models/secure_rooms/card_reader.rb (hunk 10 above)
Comment (waggles) [inline vendor/engines/secure_rooms/app/models/secure_rooms/card_reader.rb]: yeah, this PR is primarily just to spit out the "deny" so these are only here for their finders to fail
